keps/sig-cli/kubectl-apply-prune-beta.md

```diff
@@ -0,0 +1,141 @@
+
+### Goals
+
+Propose a solution for better `kubectl apply --prune` so that
+- it can find the correct set of objects to delete
+- it can avoid mess up when changing directories
+
+### Non-Goals
+
+- Make `--prune` default on
+- Improvement on Garbage Collections
+
+## Proposal
+
+### Challenges
+The challenges of pruning are mainly from two aspects.
+- Find out correct objects to delete. When a directory is applied after some updates, it loses the track of objects that are previously applied from the same directory.
+- Delete objects at proper time. Assume the set of objects to be deleted is available, when to delete them is another challenge. Those objects could be still needed during rolling out other objects.
+
+In this design, we focus on a solution for the first challenge. For the second challenge, we can blacklist certain types such as ConfigMaps and Secrets for deletion. Generally deleting objects at proper time can be better addressed by Garbage Collection.
+
+
+### Root Object
+To solve the first challenge, a concept of root object is proposed. A root object is a ConfigMap object that contains the GVKN information for all the objects that are to be applied. The root ConfigMap objects should be created by other tools such as kustomize, Helm.  When generating resources to be consumed by kubectl apply, those tools are responsible to generate the associated root ConfigMap object to utilize better pruning.
+
+Here is an example of root ConfigMap object. It has a predictable name or a label `kubectl.kubernetes.io/applyRoot: true` indicating that it is a root object.
+```
+apiVersion: v1
+kind: ConfigMap
+metadata:
+ name: predictable-name-foo
+ labels:
+   kubectl.kubernetes.io/applyRoot: true
+data:
+ object1:  namespace1:group1:version1:kind1:somename
+ object2:  namespace1:group2:version2:kind2:secondname
+ object3:  namespace3:group3:version3:kind3:thirdname
+```
+
+The presence of a root object indicates apply need to handle prune. When kubectl apply is given such a root object, it finds the live root object with the same name. 
+- If a live root object is not found, Kubectl will create the root object and the corresponding set of objects
+- If a live root object is found, then the set of live objects will be found by reading the data field of the live root object. Kubectl compares the live set with new set of objects and takes following steps: 
+  - set the root object data as the union of two sets and update the root object
+  - create new objects
+  - prune old objects
+  - remove the pruned objects from root object’s data field
+
+
+### Child Objects
+The objects passed to apply will have labels or annotations pointing back to the root object for the same apply.
+```
+ labels:
+    kubectl.kubernetes.io/applyOwner: v1:ConfigMap:predictable-name-foo
+```
+The presence of this label is to confirm that a live object is safe to delete when it is to be pruned.
```

Comment:
So every object must have this label to be associated with the root object and thus the tool is able to construct the new set  and old set? I guess I understand how the new set and old set are constructed.

But I still feel it will be safer that the root object must be consistent with the actual objects in the system before `--prune` is allowed to proceed. The tools creating child objects are responsible for updating the root object to make it consistent. Otherwise it's still possible for making mistakes.

Reply:
> But I still feel it will be safer that the root object must be consistent with the actual objects in the system before --prune is allowed to proceed. The tools creating child objects are responsible for updating the root object to make it consistent. Otherwise it's still possible for making mistakes.

By this proposal, Kubectl will assume the root object is consistent with the actual objects. I agree with you that the tools are responsible to make sure this is true.